### PR TITLE
Bugfix - Handling blocked users gracefully when reacting to their messages

### LINKED ISSUE #406: Autoreact on suggestions doesnt handle blocked users gracefully
## Overview

If someone blocks our TJ-Bot, it can not react on their suggestions. Right now, the code throws exceptions in that case:

log

## Proposed action

While we probably can not really work around that easily with the current system, we should at least handle this situation gracefully and not crash.

### PR DESCRIPTION
**Pull request by MaiTheLord**
warped the line that reacts to the message in a `try catch` block. see file changes.
resolves #406.

### CODE REVIEW

**Comment by MaiTheLord** [comment]
> Why complete

can't catch `queue()`, afaik.

**Comment by MaiTheLord** [comment]
iirc, `queue()` throws exceptions on the executing thread, thus you can't catch them on the calling thread.

**Comment by MaiTheLord** [comment]
```java
.queue(ignored -> {},
exception -> {
    if (exception instanceof ErrorResponseException && ((ErrorResponseException) exception).getErrorResponse() == ErrorResponse.REACTION_BLOCKED) {
        return;
    }
    exception.printStackTrace();
});
```
is this better?

**Comment by Zabuzard** [comment]
u have to use our logging system, not e.printStackTrace. so at least `logger.error("...", e)`

**Comment by Zabuzard** [comment]
> The best thing to do is to throw it, probably, but how do I do it from inside the function? 🤔

Then you would throw from within the JDA context where the lambda is executed, which is detached from our flow - not good.

If you stick with `queue`, you should log the error using the logger. If you stick with `complete`, try-catch is fine. That means that the call is blocking though and not async. Which means that you could potentially block one of our bot-core-system threads for a long time. We only have a few of these, so if there are many reactions going on and they, for whatever reason, take long to complete, that can block the whole bot and stuff like `/tag foo` can not be executed anymore and will die.

tldr, dont block our threads.

**Comment by MaiTheLord** [comment]
that should do it.

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java [hunk 1]
```diff
@@ -56,7 +58,15 @@ private static void reactWith(@NotNull String emoteName, @NotNull String fallbac
...
             return message.addReaction(fallbackUnicodeEmote);
-        }).queue();
+        }).queue(ignored -> {
+        }, exception -> {
+            if (exception instanceof ErrorResponseException && ((ErrorResponseException) exception)
+                .getErrorResponse() == ErrorResponse.REACTION_BLOCKED) {
+                return;
+            }
+
+            logger.error(exception.getMessage(), exception);
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:68]: Just log the exception, it'll print the exception message twice now

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:68]: to add to what tais says, the loggers `logger.foo(..., exception)` already appends the full error message, stack trace whatsoever to the log. the first part is just ur text of whats going on and to provide context. like `logger.error("Attempted to react to a suggestion, but failed for an unknown reason", exception);`

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:68]: leads me into thinking that u actually didnt try out the code to see how it looks, but just coded and pushed it right of the bat. u shouldnt do that, trigger ur new code to verify it.

Comment (MaiTheLord) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:68]: I am pretty sure it won't. how do you suggest I log it?

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:68]: > I am pretty sure it won't.

I am pretty sure it does. Right now it will say

> Foo bar
> FooException: Foo bar
> at ...
> at ...

So the message is twice. Just completely unecessary. Instead, add context.

> how do you suggest I log it?

Just like I showed above. So
```java
logger.error("Attempted to react to a suggestion, but failed for an unknown reason", exception);
```

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:68]: Dont assume. Try it out and you will see yourself.

Comment (MaiTheLord) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:68]: I just tested it by removing the `if` statement completely, blocking the bot and sending a message, and I got this:
https://gist.github.com/MaiTheLord/90c678e4948d733c7a4edcb7a48dd715
it sent it twice as it tries to send two embeds.

Comment (MaiTheLord) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:68]:
```
22:01:21.247 [ForkJoinPool.commonPool-worker-1] ERROR org.togetherjava.tjbot.commands.basic.SuggestionsUpDownVoter - 90001: Reaction blocked
net.dv8tion.jda.api.exceptions.ErrorResponseException: 90001: Reaction blocked
```

Comment (MaiTheLord) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:68]: When it logs an exception the stack trace shows this.

image

is that supposed to happen? it doesn't happen if I don't use any error handler for the `queue()`.

**Comment by MaiTheLord** [comment]
When it logs an exception the stack trace shows this.

image

is that supposed to happen? it doesn't happen if I don't use any error handler for the `queue()`.

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java (hunk 1 above)
Comment (MaiTheLord) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:68]: this solves it, I guess. is that good?
```java
logger.error("", exception);
```

Comment (MaiTheLord) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:68]: > leads me into thinking that u actually didnt try out the code to see how it looks, but just coded and pushed it right of the bat. u shouldnt do that, trigger ur new code to verify it.

actually I did test it. I just didn't pay much attention to the 2 first line of the stacktrace, unfortunately,

Comment (MaiTheLord) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:68]: > ```java
> logger.error("Attempted to react to a suggestion, but failed for an unknown reason", exception);
> ```

I wouldn't say "unknown reason", as the reason is stated in the stacktrace. maybe just
```java
logger.error("Attempted to react to a suggestion, but failed", exception);
```

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:68]: sounds good

**Comment by Tais993** [comment]
I don't know or it should, please give the full stacktrace

**Comment by MaiTheLord** [comment]
Here ya go
https://gist.github.com/MaiTheLord/90c678e4948d733c7a4edcb7a48dd715

**Comment by Zabuzard** [comment]
Sounds like you logged the error now. Didnt you want to ignore it and only log unknown failures? Whats the code that produced this?

**Comment by MaiTheLord** [comment]
I commented the part that ignores the error, for the sake of seeing a stacktrace, as I couldn't think of any other way to make it fail.

**Comment by MaiTheLord** [comment]
I found another way to make it fail (by revoking the bots permissions).
https://gist.github.com/MaiTheLord/99335c4206d4f81da055df4fbb5e6deb

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java (hunk 1 above)
Comment (MaiTheLord) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java]: can I resolve this?

**Comment by Zabuzard** [comment]
Not sure what your question or confusion is, to be honest. This is the expected behavior - it logs the unknown error.

**Comment by MaiTheLord** [comment]
I wasn't sure about the `caused by ContextException`.

when I look at it now, seems like its JDA's way to handle async exceptions.

**Comment by Tais993** [comment]
> When it logs an exception the stack trace shows this.

ah yeah, that's just the exception, scroll up and you see the cause.

>>> I found another way to make it fail.
https://gist.github.com/MaiTheLord/99335c4206d4f81da055df4fbb5e6deb
produced by latest code, https://github.com/Together-Java/TJ-Bot/commit/8d554a3ede924c5db838bb5219f86cc92b5008f4, and revoking the Add Reactions permission

There's a ton of ways to make it fail, but we shouldn't do anything with this.
It logs the error, the bot owner can add permissions, not our issue.
Just keep it like this

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java [hunk 2]
```diff
@@ -56,7 +58,15 @@ private static void reactWith(@NotNull String emoteName, @NotNull String fallbac
                     "Unable to vote on a suggestion with the configured emote ('{}'), using fallback instead.",
                     emoteName);
             return message.addReaction(fallbackUnicodeEmote);
-        }).queue();
+        }).queue(ignored -> {
+        }, exception -> {
+            if (exception instanceof ErrorResponseException && ((ErrorResponseException) exception)
+                .getErrorResponse() == ErrorResponse.REACTION_BLOCKED) {
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:64]:
```suggestion
            if (exception instanceof ErrorResponseException responseException &&
                responseException.getErrorResponse() == ErrorResponse.REACTION_BLOCKED) {
```

**Review by Tais993 — CHANGES_REQUESTED** [review]

**Review by Tais993 — APPROVED** [review]

**Comment by MaiTheLord** [comment]
Zabuzard, can you review please?

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java [hunk 3]
```diff
@@ -56,7 +58,15 @@ private static void reactWith(@NotNull String emoteName, @NotNull String fallbac
                     "Unable to vote on a suggestion with the configured emote ('{}'), using fallback instead.",
                     emoteName);
             return message.addReaction(fallbackUnicodeEmote);
-        }).queue();
+        }).queue(ignored -> {
+        }, exception -> {
+            if (exception instanceof ErrorResponseException responseException
+                    && responseException.getErrorResponse() == ErrorResponse.REACTION_BLOCKED) {
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:64]: Maybe add a quick comment to explain this use case:
```java
// User blocked the bot, hence the bot can not add reactions to their messages. Nothing we can do here.
return;

**Review by Zabuzard — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java (hunk 3 above)
Comment (MaiTheLord) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:64]: sure.

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java (hunk 1 above)
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java]: yes

**Review by Tais993 — APPROVED** [review]

**Review by Zabuzard — APPROVED** [review]
